llvm/include/llvm/Analysis/CmpInstAnalysis.h

```diff
@@ -95,24 +95,28 @@ namespace llvm {
   /// Represents the operation icmp (X & Mask) pred C, where pred can only be
   /// eq or ne.
   struct DecomposedBitTest {
-    Value *X;
+    Value *X = nullptr;
     CmpInst::Predicate Pred;
     APInt Mask;
     APInt C;
   };
 
   /// Decompose an icmp into the form ((X & Mask) pred C) if possible.
-  /// Unless \p AllowNonZeroC is true, C will always be 0.
+  /// Unless \p AllowNonZeroC is true, C will always be 0. If \p
+  /// DecomposeBitMask is specified, then, for equality predicates, this will
```

Comment:
I recommend landing this change separately. Then we can remove duplicate logic in other places.
https://github.com/llvm/llvm-project/blob/3ed83630b263f667bacd1ef1a65f26d8665c282e/llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp#L873-L895
https://github.com/llvm/llvm-project/blob/3ed83630b263f667bacd1ef1a65f26d8665c282e/llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp#L2755-L2787
https://github.com/llvm/llvm-project/blob/3ed83630b263f667bacd1ef1a65f26d8665c282e/llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp#L3732-L3765

Reply:
BTW, I am still confused by the parameter name "DecomposeBitMask". It is just what this helper function does.
I think "MatchCanonicalForm" would be better.

Reply:
The function decomposes and finds better forms for comparisons by testing for bit conditions of the compared value -> "decomposeBitTest"

The new capability decomposes cases where we bitmask a value via `and` -> "decomposeBitMask"

Does this explanation help, or is it still too confusing?

Either way, "MatchCanonicalForm" seems to be much more vague and isn't really describing what the capability does.

Reply:
I added  https://github.com/llvm/llvm-project/pull/136367  -- we can discuss there.